Add initial commands

tomalec opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->
See project README for the overview.

- [Switch to ESM](https://github.com/tomalec/dewped/pull/5/commits/af0df0ca4fb534ba15a1875c180bd4ae2f57bc8c)
- [Port L-x command](https://github.com/tomalec/dewped/pull/5/commits/0777cb1ffcacf788434b2e62ef85e4f99f0f1794) 
   copy and adjust from 
   https://github.com/woocommerce/grow/blob/e4e0d31c0844f038d1657ca74e2bc47725bc294a/packages/js/github-actions/actions/get-plugin-releases/src/get-plugin-releases.js
   add a few basic tests
- [Update TS & lint to support ESM-importing .js files](https://github.com/tomalec/dewped/pull/5/commits/baa966145c97d5a0150de3b37f045979fea882b4)
- [Add command to check WP, WC & local dependencies' versions,](https://github.com/tomalec/dewped/pull/5/commits/53ccbacb74995294bb36d80b35f111e954da0680) 
   for a given platform version(s),
   taking the package names from CLI or externalized.json file.
- [Remove example hello commands](https://github.com/tomalec/dewped/pull/5/commits/d82228c5a30aec7d5e1670ef545bc9b2c2165122)
- [Add not-found plugin,](https://github.com/tomalec/dewped/pull/5/commits/4ddc7c5790372d34dcda2325a4a5a4e9e69540ad) 
   to improve devX for folks with dyslexia.
- [Uninstall @oclif/plugin-plugins](https://github.com/tomalec/dewped/pull/5/commits/da1a2f09575f449843cbebd68366a707a29730b1) 
   As it is not used currently.
- [Add examples to commands](https://github.com/tomalec/dewped/pull/5/commits/9f5671d5834e60affdf7dd6fe4041c936427f26d)
- [Update required node engine](https://github.com/tomalec/dewped/pull/5/commits/a181ea5d863842aada5a19ad80ac85240240d852)
- [Switch test workflow to npm](https://github.com/tomalec/dewped/pull/5/commits/1853305a7f21e6d1cbd2d8ffd4251371d08835cf)
- [Run npm audit fix and bump some deps](https://github.com/tomalec/dewped/pull/5/commits/16aee32f59bb666a18a81a0f6c21b05b61712a5c)

### Screenshots:
![image](https://github.com/tomalec/dewped/assets/17435/d71e2c17-895a-42d9-9504-74675ddfe052)

![image](https://github.com/tomalec/dewped/assets/17435/2d2d1e43-2486-4d5a-8dab-4a2a211f502c)



### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

1. Run automated tests
   ```sh
   npm i && npm run build && npm run test
   ```
2. Link npm package for testing `npm link`
3. Go to an extension repo directory, like AutomateWoo
4. Use the link `npm link dewped`
5. Test the commands
```
dewped help
dewped l-x woocommerce 4
dewped pdep --wpVersion=6.0.1 -c=7.8.1 -d=admin/assets/.externalized-admin.json 
dewped pdep --wpVersion=6.0.1 -c=7.8.1 -d=admin/assets/.externalized-admin.json --json

```



### Additional details:

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation
If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.
Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

tomalec committed:
Switch to ESM

tomalec committed:
Port L-x command

copy and adjust from https://github.com/woocommerce/grow/blob/e4e0d31c0844f038d1657ca74e2bc47725bc294a/packages/js/github-actions/actions/get-plugin-releases/src/get-plugin-releases.js

add a few basic tests

tomalec committed:
Remove example hello commands

tomalec committed:
Update TS & lint to support ESM-importing .js files

tomalec committed:
Add command to check WP, WC & local dependencies' versions,

for a given platform version(s),
taking the package names from CLI or externalized.json file.

tomalec committed:
Add not-found plugin,

to improve devX for folks with dyslexia.

tomalec committed:
Uninstall `@oclif/plugin-plugins`

As it is not used currently.

tomalec committed:
Add examples to commands

tomalec committed:
Regenerate readme

tomalec committed:
Remove duplicated docs from aliases

tomalec committed:
Update required node engine

tomalec committed:
Update README descriptions,

remove aliases from commands list.

tomalec committed:
Switch test workflow to npm

tomalec committed:
Add package-lock.json to the repo

tomalec committed:
Fix typo in the pdep command name

tomalec committed:
Run npm audit fix and bump some deps

eason9487 commented on package-lock.json:
There are some file changes after running `npm install`. Maybe this file is not up-to-date with the package.json

eason9487 commented on bin/run.js:
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env node
+// eslint-disable-next-line node/shebang
```
Looks like this ignore is not needed. BTW, Is it expected not to include .js files in `npm run lint` script?

tomalec replied:
It seems we actually need this, as for some reason, on Windows builds, it throws here https://github.com/woocommerce/dewped/actions/runs/5520307380/jobs/10066744858

tomalec replied:
I've overridden the check in https://github.com/woocommerce/dewped/pull/5/commits/3eae059a846f6a60eac0b41005d4b594e514c8ba

eason9487 commented on README.md:
The description was not added in this PR but I wonder if the `dewped help` supports querying multiple commands at once?

https://github.com/woocommerce/dewped/blob/16aee32f59bb666a18a81a0f6c21b05b61712a5c/README.md?plain=1#L30-L39

I got the same output with `dewped help l-x pdep` and `dewped help l-x`.

tomalec replied:
:thinking: "command**s**" argument name is really confusing, but the description states it in the singular form "COMMANDS  Command to show help for. ".

The problem is the source of that description comes from oclif repo. Then we autogenerate the docs, so we would have to keep on fixing it after each re-generation :/ 
I made a PR there https://github.com/oclif/plugin-help/pull/520. Let's see how it flows. If it will not get traction in a few weeks and will still bother us, I'd change it and start thinking of another sustainable solution.

eason9487 commented on README.md:
```diff
@@ -100,236 +47,112 @@ DESCRIPTION
 
 ARGUMENTS
-  PLUGIN  [default: .] Plugin to inspect.
+  SLUG    [default: wordpress] Slug of the extension
```
This command only supports extensions that have been published to https://wordpress.org/. Other extensions, for example, AutomateWoo, are not supported.

It's recommended to describe the supported scope here, and perhaps also the slug parameter in `src/commands/l-x.ts`

eason9487 reviewed (commented):
I got an error when running `dewped help` with Google Listings & Ads repo. This error won't happen with Node.js >= 16. Looks like the build commands cannot be run with Node.js < 16?

```
(node:13121) SyntaxError Plugin: dewped: The requested module 'node:fs/promises' does not provide an export named 'constants'
module: @oclif/core@2.8.11
task: toCached
plugin: dewped
```

eason9487 commented on src/commands/platform-dependency-version.ts:
```diff
@@ -0,0 +1,161 @@
+     Name                  WordPress 6.0.3 WooCommerce  Local
+     ───────────────────── ─────────────── ──────────── ─────
+     @wordpress/components 19.8.5                             `,
+    `<%= config.bin %> pdep -w=6.0.3 -c=7.0.1 @wordpress/components -d=.externalized.json
```
Curious about what the `@wordpress/components` in this example is supposed to represent?

tomalec replied:
For the first example `--wpVersion=6.0.3 @wordpress/components`:

Let's say you're using/about to use `import { Flex } from '@wordpress/components'`. Then you would like to check what version of that package to anticipate when WordPress 6.0.3 is running. So you could change the version of require or test up to version, use that version for automated unit tests, check which version to debug, etc.
Then you can run `dewped --wpVersion=6.0.3 @wordpress/components`.

For the second one `pdep -w=6.0.3 -c=7.0.1 @wordpress/components -d=.externalized.json`:
You would like to check versions for all packages that are externalized and reported in the `.externalized.json`. The `@wordpress/components` in the example is to show-case that if `-d` is provided, other packages from CLI are ignored.

eason9487 commented on bin/dev.js:
I tried to run this script directly through `./bin/dev.js` but it failed:
```
env: ts-node: No such file or directory
```
I guess it needs to install a global `ts-node` for running this script in that way.

Although devs can use `./node_modules/.bin/ts-node ./bin/dev.js` to run the dev mode, considering `ts-node` is listed in the `devDependencies` of package.js already, could it also add a script to package.json to run `ts-node ./bin/dev.js`? So that it should be easier to dev with the `ts-node` installed in the local `node_modules`.

tomalec replied:
Addressed in https://github.com/woocommerce/dewped/pull/5/commits/9ae725cc0971db000f953bba73889399bc28f986

eason9487 commented on src/commands/platform-dependency-version.ts:
```diff
@@ -0,0 +1,161 @@
+        name: dep,
+        wpVersion: wpDependencies[dep] || null,
+        wcVersion: wcDependencies.get(dep) || null,
+        localVersion: localDependencies[dep] || null,
```
It would occur a type error if run this command under a directory without a package.json file. Surprisingly, TypeScript didn't even warn about this.

tomalec replied:
Thanks for finding that :) 
Fixed in https://github.com/woocommerce/dewped/pull/5/commits/a49fc0fc6f22ceff4468881fc4744dfb1715b28d

And slightly adjusted the behavior in https://github.com/woocommerce/dewped/pull/5/commits/3332851991534df274ca5b7c00cb08fd2a16d719

eason9487 commented on src/commands/platform-dependency-version.ts:
```diff
@@ -0,0 +1,161 @@
+     react                   17.0.2                                     `,
+  ]
+
+  public static enableJsonFlag = true
```
Maybe this flag could be disabled as the `--output` also has JSON format and will conflict with this flag.

tomalec replied:
From what I tested 
![image](https://github.com/woocommerce/dewped/assets/17435/e6b6f9b8-78f2-4abc-84aa-b13c9de818d7)
I'd not call it a "conflict" but rather a "precedence", or do you mean some other result?

I agree it's a bit redundant to have `--json` and `--output=json`.

What I like about `--json` is that it gives the coloring, plus it makes every result a JSON, like errors as well.
![image](https://github.com/woocommerce/dewped/assets/17435/dfaf6914-9f1f-4382-a40a-c1fd72a6f02d)

The way I reason about it is that `--output` is to format only the resulting table, while `--json` is to force any command output to JSON.

eason9487 replied:
The main reason is that the two flags seem to be in conflict when a command user is trying to understand how to use this command.

![2023-07-12 11 43 42](https://github.com/woocommerce/dewped/assets/17420811/5770ea8f-bed7-4924-857d-bdd324aed11c)

It's necessary to actually run it to understand the precedence relationship between them. While they don't cause problems with usage, perhaps a brief mention of their precedence in the `--help` result would be nice.

eason9487 commented on test/commands/l-x.test.ts:
```diff
@@ -0,0 +1,58 @@
+  .command(['l-x', 'google-listings-and-ads'])
+  .it('when slug (first) argument is provided logs that it\'s "Fetching L-2 versions from {slug}"', ctx => {
+    expect(ctx.stdout).to.contain('Fetching L-2 versions google-listings-and-ads!')
+  })
```
Looks like the version offset was missing in this test?

tomalec replied:
Addressed in https://github.com/woocommerce/dewped/pull/5/commits/4a8d53b7ffb9ddeed28c15cb6fa82b11a1fe32b5

eason9487 commented on test/commands/platform-dependency-version.test.ts:
```diff
@@ -0,0 +1,145 @@
+    .stdout()
+    .command(['platform-dependency-version', '--wpVersion=6.0.3', '@oclif/core'])
+    .it('render it in the table', ctx => {
+    // Calcute the table indention.
```
The indentation looks incorrect and the Calcute seems to be a typo.

tomalec replied:
Addressed in https://github.com/woocommerce/dewped/pull/5/commits/a6e324315278981a3929a692eeb573e073b18ec2

tomalec committed:
Merge remote-tracking branch 'origin/trunk' into add/initial-commands

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1255603372

eason9487 commented on test/commands/platform-dependency-version.test.ts:
```diff
@@ -0,0 +1,145 @@
+    })
+  })
+
+  describe('if there is a package.json in working directory, it checks the dependency versions there', async () => {
```
All tests under this `describe` block were not performed. It seems to be mainly affected by the `async` syntax.

tomalec replied:
Addressed in https://github.com/woocommerce/dewped/pull/5/commits/b9b142c9b694ea8e86fd7b5dbd9f5a1522122d20

eason9487 commented on test/commands/platform-dependency-version.test.ts:
```diff
@@ -0,0 +1,145 @@
+      .stdout()
+      .command(['platform-dependency-version', '--wpVersion=6.0.3', '--wcVersion=7.8.1', '@woocommerce/admin-layout', '--wcDEWP=7.0.1'])
+      .it('which may miss some packages', ctx => {
+        // Prints packages with their versions.
```
This comment doesn't match the test.

tomalec replied:
Addressed in https://github.com/woocommerce/dewped/pull/5/commits/ec1a8cdf26905cb1e4cd84da87ca96521b932a3a

eason9487 reviewed (commented):
Thanks for the work.

Found some issues that may need to be addressed. Other comments about docs or repo settings are minor things but also recommend adjusting a bit.

tomalec committed:
Use `node:fs.constants` to support Node 14

Replace `node:fs/promises.constants`,
addresses https://github.com/woocommerce/dewped/pull/5#pullrequestreview-1518513225

tomalec committed:
Start linting js files, remove non-needed comment

addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1255607077

tomalec committed:
Note that slug needs to be in WPORG,

mention the source of l-x clone.

Addresses https://github.com/woocommerce/dewped/pull/5/files/16aee32f59bb666a18a81a0f6c21b05b61712a5c#r1255646696

tomalec committed:
Add `dev` script to run local ts-node

Addresses https://github.com/woocommerce/dewped/pull/5/files/16aee32f59bb666a18a81a0f6c21b05b61712a5c#r1257913794

tomalec committed:
Disable node/shebang check for `run.js`

to silence false positives for Windows machines.

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1259742229

tomalec committed:
Prevent TypeError if pockage.json is missing

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1258025460

tomalec committed:
Don't show column headers if there is no source of data

tomalec committed:
Do not include WC & local versions if not asked

tomalec committed:
Fix a typo in code comment

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1258061792

tomalec committed:
Fix code comments to match checking the presence or the lack of package.

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1258086035

tomalec committed:
Fix running async describe

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1258082842

tomalec committed:
Add the actual test for offset in l-x

Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1258051418

tomalec commented:
Thank you, @eason9487 for the detailed review!

I addressed/answered all your comments. Would you mind taking another look?

I also changed the behavior a bit.
Now if you don't provide the WC version, or if there is no `package.json` the output data and column will not contain respective columns/properties. IMO it makes the output shorter & easier to read, plus gives an insight into what was checked and what was not.

eason9487 reviewed (approved):
Thanks for the adjustments. Tested with some argument/flag combinations of commands and they worked well. LGTM.

Left a [further discussion](https://github.com/woocommerce/dewped/pull/5#discussion_r1260536095) about adjusting the description for the output format flags in the `platform-dependency-version` command, and it's not a blocker.

tomalec committed:
Adjust `--json` flag description

 to mention its precedence over the `--output` flag.

 Addresses https://github.com/woocommerce/dewped/pull/5#discussion_r1260536095
